### PR TITLE
add failing unit test for listbox issue #1936

### PR DESCRIPTION
**Pull request by donandren**
- What does the pull request do?
Emulates issue #1936
- What is the current behavior?
- What is the updated/expected behavior with this PR?
 Outline the issue in unit test and help find root cause of the issue
- How was the solution implemented (if it's not obvious)?
  Emulated user actions that are proven show the problem in the most easy way that's found so far

Checklist:

- [x] Added unit tests (if possible)?

Emulates issue #1936

### CODE REVIEW

**Comment by akunchev** [comment]
The problem look like is in ItemVirtualizerSimple.RecycleContainers:
https://github.com/AvaloniaUI/Avalonia/blob/d828957d6e7620fbee9cac64225d8cf42d058ccf/src/Avalonia.Controls/Presenters/ItemVirtualizerSimple.cs#L368

`...
var containers = generator.Containers.ToList();
...
`
There is no guarantee Containers are ordered by the index of visualization and it's bit unpredictable in specific cases which container will be updated which not.

if we change this line to:

`...
var containers = generator.Containers.OrderBy(c=>c.Index).ToList();
...
`
this bug is fixed, but i have no idea is this the most correct, performant and elegant solution.
I personally don't like using linq in this part of the code, so this doesn't look like the proper solution for me.
if i can have some feedback @grokys @jkoritzinsky  i can work on fixing the issue in a proper way.

**Comment by grokys** [comment]
I've not had time to look at the failing test properly, but the point about the containers not being ordered sounds about right.

You're also right that `OrderBy` probably isn't a good idea here. It would probably make more sense to ensure that the containers are inserted in the list in the correct order?

File: tests/Avalonia.Controls.UnitTests/ListBoxTests.cs [hunk 1]
```diff
@@ -170,6 +171,109 @@ public void ScrollViewer_Should_Have_Correct_Extent_And_Viewport()
...
+        [Theory]
+        [InlineData(ItemVirtualizationMode.Simple)]
+        [InlineData(ItemVirtualizationMode.None)]
+        public void When_Added_Removed_AfterItems_Reset_Should_Work(ItemVirtualizationMode virtMode)
+        {
+            using (UnitTestApplication.Start(TestServices.StyledWindow))
+            {
+                var items = new ObservableCollection<string>();
+
+                Action create = () =>
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/ListBoxTests.cs]: Could you turn these lambdas into local functions? It's a bit easier to read for me that way.

Comment (akunchev) [reply tests/Avalonia.Controls.UnitTests/ListBoxTests.cs]: @grokys no more lambdas now

**Comment by grokys** [comment]
Ok, I've taken a look at this, and it seems that the test has a lot of stuff in it that isn't actually needed. Are you sure that everything that's there is necessary @donandren ?

In addition, to fix it, it looks like just changing `_containers` to be a `SortedDictionary` solves the problem.

If you don't mind I'm going to refactor the test to just contain the necessary parts and commit a fix. Is that ok?

**Comment by grokys** [comment]
OK @donandren simplified the test and pushed a fix. Could you take a look and let me know if I've missed anything?

**Comment by akunchev** [comment]
all LGTM, test was more complicated as i haven't found the root of the reason properly to simplify it myself. thanks, great work.

**Comment by donandren** [comment]
@grokys all LGTM, test was more complicated as i haven't found the root of the reason properly to simplify it myself. thanks, great work.

**Pull request merged**